### PR TITLE
Forge/changes

### PR DESCRIPTION
**Pull request by matthewpeterkort**
Adds special cases functionality / commands mostly for edge cases / dev-ops purposes. Existing 1 profile 1 remote use case should still work as intended. 

git-drs push [optional remote] default 'origin' - push files / drs objects with multi instance support. Looks in .drs/lfs/objects and pushes all drs objects to the new server. Files that do not exists locally will be skipped as it is assumed that records that have been fetched from another instance have resolvable data in whatever bucket they were pointing to on the other instance

Note: This git-drs push use case should only be needed for users with more than 1 profile. Standard use case using the transfer client is still the main use case.

git-drs fetch [optional remote] default 'origin' - fetches/gets all drs records from remote and updates
.drs/lfs/objects to reflect what is on the remote. Does not delete records that did not exist in remote

Note: This is another special cases use case for fetching indexd records from a specified remote. The average user probably doesn't need this.

git-drs precommit [optional remote] default '
fetches all drs records that exists in the .git/lfs land and updates .drs/lfs/objects. Does not delete records that do not exist on local. This already exists for the most part. I don't think I have changed this functionality.

Note: precommit will eventually get abstracted into a pre-push command at some point.

"git-lfs pull local -I META/Directory.ndjson" works as expected already. No need for a git-drs pull command.
you simply have to specify what remote you want to pull from.

Warning: When you run this the logging is pretty aggressive since the transfer client isn't getting as much use as it normally does. 
```
#!/bin/bash
set -e
set -x  # Print each command as it's executed

# Set repo name and remote URL
REPO_NAME="git-drs-e2e-test"
GIT_USER="cbds"
# REMOTE_URL="https://source.ohsu.edu/$GIT_USER/$REPO_NAME"
REMOTE_URL="git@source.ohsu.edu:$GIT_USER/$REPO_NAME.git"

# Clean up if rerunning (don't fail if not removable)
rm -rf "$REPO_NAME" || true

# Create directory and initialize git
mkdir "$REPO_NAME"
cd "$REPO_NAME"


# Step 1: Initialize Repository
git init
git drs init
git drs remote add gen3 local --url https://caliper-training.ohsu.edu --cred ~/.gen3/credentials.json --project cbds-git_drs_test --bucket cbds


# set branch / add remote
git branch -M main
git remote add local "$REMOTE_URL"

git lfs track "*.greeting"
git add .gitattributes

# Step 2: Create and Commit Initial Files
mkdir -p data/A data/B data/C

echo "hello $(date)" > data/A/simple.greeting
sleep 1
echo "hello $(date)" > data/B/simple.greeting
sleep 1
echo "hello $(date)" > data/C/simple.greeting

git add data/
git add .drs/config.yaml

git drs precommit local
git commit -m "Initial commit: Add .greeting files with 'hello'"

# Prompt user for remote if not set
git push --set-upstream local main


# Step 3: Clone the repository for the second profile
git clone "$REMOTE_URL"
cd "$REPO_NAME"

# Initialize git-drs in the new clone
git drs init

# Add the same Gen3 remotes so the second profile knows where to get/send data
# The goal is to FETCH from 'local' and PUSH to 'dev'
git drs remote add gen3 local --url https://caliper-training.ohsu.edu --cred ~/.gen3/credentials.json --project cbds-git_drs_test --bucket cbds
git drs remote add gen3 dev --url https://calypr-dev.ohsu.edu --cred ~/.gen3/credentials-dev.json --project cbds-git_drs_test --bucket cbds

# Step 4: Fetch DRS records from the first remote ('local')
echo "Attempting to fetch DRS records from 'local'..."
# The custom command is: git-drs fetch [optional remote] default 'origin'
# This pulls the DRS records from the Gen3 'local' instance and updates .drs/lfs/objects
git drs fetch local

# Verify the local DRS index has been updated by checking the contents of .drs/lfs/objects
if [ -z "$(ls -A .drs/lfs/objects/)" ]; then
    echo "ERROR: Fetch failed, .drs/lfs/objects is empty."
    exit 1
fi
echo "Successfully fetched DRS records from 'local'."

echo "Attempting to pull data/A/simple.greeting from 'dev' (Expected: Success)"
git remote add dev "$REMOTE_URL"

git drs push dev 
# The pull command will try to resolve the DRS ID for the file and download the object
git-lfs pull dev -I data/A/simple.greeting

# Check the exit status of the previous command
if [ $? -eq 0 ]; then
    echo "✅ SUCCESS: data/A/simple.greeting downloaded correctly."
    # Optionally verify file content and size if needed
    if [ -f "data/A/simple.greeting" ]; then
        echo "File data/A/simple.greeting exists locally."
    else
        echo "❌ ERROR: File data/A/simple.greeting not found locally after pull."
        exit 1
    fi
else
    echo "❌ FATAL ERROR: data/A/simple.greeting failed to download when it was expected to succeed."
    exit 1
fi

# --- Cleanup (Delete all DRS records) ---
echo "Running cleanup command: Deleting all DRS records for project 'cbds-git_drs_test' on 'dev'..."
git-drs delete-project cbds-git_drs_test --remote dev
# Note: The 'git-drs delete-project' command itself should exit successfully (0) if the operation was initiated.
# The side effect is that the DRS records are now gone from the 'dev' Gen3 instance.

# --- Test Download of Second File (Expected: FAILURE / ERROR) ---
echo "Attempting to pull data/B/simple.greeting from 'dev' (Expected: Failure/Error)"

# The pull command should now fail because the DRS record was deleted by the 'delete-project' command.
git-lfs pull dev -I data/B/simple.greeting

# Check the exit status of the previous command
if [ $? -ne 0 ]; then
    echo "✅ SUCCESS: Pull of data/B/simple.greeting failed as expected (non-zero exit code)."
    # The actual error message (like "error: Fetching reference..." or "Could not resolve data...") 
    # will be printed by the git-lfs pull command itself.
else
    echo "❌ FATAL ERROR: data/B/simple.greeting pulled successfully, but was expected to fail after delete-project."
    # Fail the test because the cleanup step did not prevent the download.
    exit 1
fi

echo "Successfully pushed data objects to 'dev' and verified download/delete behavior."
```

### CODE REVIEW

File: client/indexd/gen3_remote.go [hunk 1]
```diff
@@ -11,8 +12,6 @@ type Gen3Remote struct {
 	Endpoint  string `yaml:"endpoint"`
 	ProjectID string `yaml:"project_id"`
 	Bucket    string `yaml:"bucket"`
-	APIKey    string `yaml:"api_key"`
-	KeyID     string `yaml:"key_id"`
```
Comment (quinnwai) [inline client/indexd/gen3_remote.go]: nice

File: client/indexd/indexd_client.go [hunk 2]
```diff
@@ -67,9 +65,26 @@ func (cl *IndexDClient) GetProjectId() string {
...
 }
 
+func (cl *IndexDClient) DeleteRecordsByProject(projectId string) error {
+	recs, err := cl.ListObjectsByProject(projectId)
+	if err != nil {
+		return err
+	}
+	for rec := range recs {
+		for _, sum := range rec.Object.Checksums {
+			err := cl.DeleteRecord(sum.Checksum)
```
Comment (quinnwai) [inline client/indexd/indexd_client.go]: I implemented the delete kinda gross where, if there's multiple records in a project with the same hash, it only deletes the first one. With our UUID system now this is fine tho cause 1 hash = 1 UUID. Just an FYI

File: client/indexd/indexd_client.go [hunk 3]
```diff
@@ -233,26 +258,30 @@ func (cl *IndexDClient) RegisterFile(oid string) (*drs.DRSObject, error) {
 	//  * addresses edge case where user X registering in project A has access to record in project B
 	//  * but still needs create a new record to so user Y reading the file in project A can access it
 	//  * even if they don't have access to project B
-	matchingRecord, err := drsmap.FindMatchingRecord(records, cl.ProjectId)
-	if err != nil {
-		return nil, fmt.Errorf("Error finding matching record for project %s: %v", cl.ProjectId, err)
+
+	var drsObject *drs.DRSObject
+	if len(records) > 0 {
```
Comment (quinnwai) [inline client/indexd/indexd_client.go]: what was the error that made you rework this?

File: client/interface.go [hunk 4]
```diff
@@ -29,12 +29,15 @@ type DRSClient interface {
...
 	// no corresponding DRS endpoint exists, so this is custom code
-	GetObjectsByHash(hashType string, hash string) ([]drs.DRSObject, error)
+	GetObjectsByHash(hash ...*drs.Checksum) ([][]drs.DRSObject, error)
 
 	///////////////////////
 	// DRS WRITE METHODS //
 	///////////////////////
 
+	// Delete all indexd records in a given project
+	DeleteRecordsByProject(project string) error
```
Comment (quinnwai) [inline client/interface.go]: a note that DRS doesn't really have the concept of a project, they do have authorizations tho and that's how we carry project / authz when translating between drs and indexd. That's a can I'm willing to kick down the road tho lol

File: client/interface.go [hunk 5]
```diff
@@ -29,12 +29,15 @@ type DRSClient interface {
 
 	// given a hash, get the objects describing it
 	// no corresponding DRS endpoint exists, so this is custom code
-	GetObjectsByHash(hashType string, hash string) ([]drs.DRSObject, error)
+	GetObjectsByHash(hash ...*drs.Checksum) ([][]drs.DRSObject, error)
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply client/interface.go]: so that you can request multiple hashshes and get back multiple drs objects. Essentially changes a single request func to a multi request func

File: client/indexd/indexd_client.go [hunk 6]
```diff
@@ -529,57 +584,64 @@ func (cl *IndexDClient) DeleteIndexdRecord(did string) error {
 }
 
 // implements /index/index?hash={hashType}:{hash} GET
-func (cl *IndexDClient) GetObjectsByHash(hashType string, hash string) ([]drs.DRSObject, error) {
+func (cl *IndexDClient) GetObjectsByHash(args ...*drs.Checksum) ([][]drs.DRSObject, error) {
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply client/indexd/indexd_client.go]: see above comment

File: client/interface.go (hunk 4 above)
Comment (matthewpeterkort) [reply client/interface.go]: Yeah I noticed that. I had deleted everything that ha a project label but then looking up individual sha's unearthed records that also had that tag. This is why I implemented the multi sha get function. Because I think that it will be better to pass in a list of sha's and get the information that way so as to not rely on a project id filter that is not the actual ground truth of what is in the db.

File: client/indexd/add_url.go [hunk 7]
```diff
@@ -25,17 +25,22 @@ import (
...
 	// get all buckets
 	baseURL := inc.Base
 	baseURL.Path = filepath.Join(baseURL.Path, "user/data/buckets")
 	// Use the AuthHandler pattern for cleaner auth handling
-	return GetBucketDetailsWithAuth(ctx, bucket, baseURL.String(), inc.Remote, &RealAuthHandler{}, httpClient)
+
+	rA, ok := inc.AuthHandler.(*RealAuthHandler)
+	if !ok {
+		return nil, fmt.Errorf("inc.AuthHandler is not RealAuthHandler")
+	}
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply client/indexd/add_url.go]: yeah that's alot cleaner. sure push it u

**Comment by matthewpeterkort** [comment]
> Also in the remote, wondering if we should have a default. Similar in git there's a git push and a git push origin main which will do the same thing if origin is the only remote specified 

Yeah I agree. Ideally the user never specifies a remote and origin is used as the default value everywhere where a remote is not specified since most users will likely not be juggling between calypr instances. I will rework the commands to reflect this.

File: client/interface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply client/interface.go]: yeah it does do batched calls. Through variadic args. you pass in an array like []*drs.Checksum... and it will accept a full batch of args. Perhaps the double array is not needed if the grouping of drs objects is not necessary then the function could continue to return []drs.DRSObject like before.

File: drsmap/drs_map.go
Comment (quinnwai) [inline drsmap/drs_map.go]: Since you're planning to use this for forge upstream, do you wanna make the DrsUUID changes here that we finalized in #79? Happy to do it in #97 when I rebase it just wanna flag that for consideration.

File: drsmap/drs_map.go [hunk 8]
```diff
@@ -34,11 +35,49 @@ type LfsFileInfo struct {
...
+	if err != nil {
+		return err
+	}
+	writtenObjs := 0
+	for drsObj := range objChan {
+		sumMap := hash.ConvertHashInfoToMap(drsObj.Object.Checksums)
+		if len(sumMap) == 0 {
+			return fmt.Errorf("Error: drs Object '%s' does not contain a checksum", drsObj.Object.Id)
+		}
+		var drsObjPath, Oid string = "", ""
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply drsmap/drs_map.go]: yeah good point no need for caps there

File: client/indexd/indexd_client.go [hunk 9]
```diff
@@ -529,59 +566,119 @@ func (cl *IndexDClient) DeleteIndexdRecord(did string) error {
...
+			continue
+		}
+
+		found := false
+		for i, rec := range records.Records {
+			if rec.Hashes.SHA256 != "" {
+				found = true
+				outobjs[rec.Hashes.SHA256] = indexdRecordToDrsObject(records.Records[i].ToIndexdRecord())
+			}
+		}
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply client/indexd/indexd_client.go]: Yeah this is by design. Added a warning comment for those who are not aware

File: cmd/query/main.go [hunk 10]
```diff
@@ -3,30 +3,32 @@ package query
 import (
 	"encoding/json"
 
-	"github.com/calypr/git-drs/config"
+	conf "github.com/calypr/git-drs/config"
 	"github.com/calypr/git-drs/drslog"
 	"github.com/spf13/cobra"
 )
 
+var remote string
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/query/main.go]: fixed in recent commit

File: cmd/addref/add-ref.go [hunk 11]
```diff
@@ -6,20 +6,24 @@ import (
...
 	Long:  "Add a reference to an existing DRS object, eg passing a DRS URI from AnVIL. Requires that the sha256 of the file is already in the cache",
 	Args:  cobra.ExactArgs(2),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		drsUri := args[0]
-		dstPath := args[1]
+		if remote == "" {
+			remote = config.ORIGIN
+		}
+		drsUri := args[1]
+		dstPath := args[2]
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/addref/add-ref.go]: fixed in most recent commit

File: cmd/remote/add/gen3.go [hunk 12]
```diff
@@ -42,108 +41,84 @@ var Gen3Cmd = &cobra.Command{
...
-		err = jwt.UpdateConfig(cred)
-		if err != nil {
-			errStr := fmt.Sprintf("[ERROR] unable to configure your gen3 remote: %v", err)
-			if strings.Contains(errStr, "apiendpoint") {
-				errStr += " If you are accessing an internal website, make sure you are connected to the internal network."
-			}
-			return fmt.Errorf("%s", errStr)
-		}
 
+	if err := jwt.UpdateConfig(log, cred); err != nil {
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/remote/add/gen3.go]: naw you wrong

File: cmd/precommit/main.go [hunk 13]
```diff
@@ -22,9 +21,13 @@ var Cmd = &cobra.Command{
...
 	Short: "pre-commit hook to create DRS objects",
 	Long:  "Pre-commit hook that creates and commits a DRS object to the repo for every LFS file committed",
-	Args:  cobra.ExactArgs(0),
+	Args:  cobra.RangeArgs(0, 1),
 	RunE: func(cmd *cobra.Command, args []string) error {
 		myLogger := drslog.GetLogger()
+		var remote config.Remote = config.ORIGIN
+		if len(args) > 0 {
+			remote = config.Remote(args[0])
+		}
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/precommit/main.go]: yep good find. Got rid of package lever var

File: cmd/list/main.go [hunk 14]
```diff
@@ -104,17 +109,20 @@ var ListProjectCmd = &cobra.Command{
...
 			return fmt.Errorf("error loading config: %v", err)
 		}
 
-		client, err := conf.GetCurrentRemoteClient(logger)
+		client, err := conf.GetRemoteClient(config.Remote(remote), logger)
 		if err != nil {
 			return err
 		}
-		objChan, err := client.ListObjectsByProject(args[0])
+		objChan, err := client.ListObjectsByProject(args[1])
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/list/main.go]: fixed in recent commit

File: cmd/transferref/main.go [hunk 15]
```diff
@@ -25,9 +25,10 @@ var (
 )
 
 var Cmd = &cobra.Command{
-	Use:   "transfer-ref",
+	Use:   "transfer-ref [remote]",
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/transferref/main.go]: fixed in most recent commit

File: client/indexd/tests/client_read_test.go [hunk 16]
```diff
@@ -204,7 +204,6 @@ func TestIndexdClient_GetDownloadURL(t *testing.T) {
 
 			client := &indexd_client.IndexDClient{
 				Base:        parseURL(mockServer.URL()),
-				Remote:      "test-remote",
```
Comment (quinnwai) [inline client/indexd/tests/client_read_test.go]: A couple unit tests failing via `go test ./...`:

```
--- FAIL: TestIndexdClient_GetDownloadURL (0.00s)
    --- FAIL: TestIndexdClient_GetDownloadURL/successful_download_URL_retrieval (0.00s)
    --- FAIL: TestIndexdClient_GetDownloadURL/successful_download_with_matching_project (0.00s)
    --- FAIL: TestIndexdClient_GetDownloadURL/successful_second_download_URL_with_different_hash (0.00s)
--- FAIL: TestIndexdClient_RegisterRecord (0.00s)
--- FAIL: TestIndexdClient_RegisterIndexdRecord_CreatesNewRecord (0.00s)
--- FAIL: TestIndexdClient_UpdateIndexdRecord_AppendsURLs (0.00s)
--- FAIL: TestIndexdClient_UpdateIndexdRecord_Idempotent (0.00s)
```

Don't worry about the add-url tests I'll fix those later

File: drsmap/drs_map.go
Comment (matthewpeterkort) [reply drsmap/drs_map.go]: fixed

File: client/indexd/tests/client_read_test.go (hunk 16 above)
Comment (matthewpeterkort) [reply client/indexd/tests/client_read_test.go]: this should be fixed now

**Review by quinnwai — CHANGES_REQUESTED** [review]
A couple minor things, one edge case to test out.

Otherwise, hash unit and integration tests works as expected!

File: cmd/precommit/main.go [hunk 17]
```diff
@@ -42,7 +42,7 @@ var Cmd = &cobra.Command{
 		myLogger.Printf("Current server: %s", dc.ProjectId)
 
 		myLogger.Printf("Preparing DRS objects for commit...\n")
-		err = drsmap.UpdateDrsObjects(string(remote), cli, myLogger)
+		err = drsmap.UpdateDrsObjects(cfg.GetRemote(remote).GetProjectId(), string(remote), cli, myLogger)
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply cmd/precommit/main.go]: done

File: drsmap/drs_map.go [hunk 18]
```diff
@@ -239,9 +254,9 @@ func writeDrsObj(drsObj *drs.DRSObject, oid string, drsObjPath string) error {
 	return nil
 }
 
-func DrsUUID(repoName string, hash string) string {
+func DrsUUID(projectId string, hash string) string {
 	// FIXME: use different UUID method? Used same method as g3t
-	hashStr := fmt.Sprintf("%s:%s", repoName, hash)
+	hashStr := fmt.Sprintf("%s:%s", projectId, hash)
 	return uuid.NewSHA1(uuid.NameSpaceURL, []byte(hashStr)).String()
```
[the remark this answers is not in the document]
Comment (matthewpeterkort) [reply drsmap/drs_map.go]: done

File: drsmap/drs_map.go [hunk 19]
```diff
@@ -34,14 +35,111 @@ type LfsFileInfo struct {
...
+	}
+
+	for drsObjKey := range outobjs {
+		val, ok := drsLfsObjs[drsObjKey]
+		if !ok {
+			myLogger.Printf("Drs record not found in sha256 map %s", drsObjKey)
+		}
+		if _, statErr := os.Stat(val.Name); os.IsNotExist(statErr) {
+			myLogger.Printf("Error: Object record found locally, but file does not exist locally. Registering Record %s", val.Name)
+			_, err = drsClient.RegisterRecord(val)
```
[the remark this answers is not in the document]
Comment (quinnwai) [reply drsmap/drs_map.go]: Okay that's fine then. I imagine we will run into this after a couple rounds of sending stuff back and forth between dev. So long as we know that isn't supported, see #101
